app/Repository/Member/MemberMatch.php

```diff
@@ -51,11 +62,259 @@ class MemberMatch {
+	 *
+	 * @param Member[] $matches
+	 * @param string $firstName
+	 */
+	private static function selectByFirstName( array &$matches, string $firstName ) {
+		// remove matches with different first name
+		foreach ( $matches as $idx => &$match ) {
+			if ( $match->firstName->getValue() ) {
+				if ( ! self::isShortNameOf( $firstName, $match->firstName->getValue() )
+				     && ! self::isShortNameOf( $match->firstName->getValue(), $firstName ) ) {
+					unset( $matches[ $idx ] );
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Compare names (case insensitive). If they are not equal, test if the
+	 * additional characters are separated either by a hyphen or a space.
+	 *
+	 * @param string $shortName
+	 * @param string $fullName
+	 *
+	 * @return bool
+	 */
+	private static function isShortNameOf( string $shortName, string $fullName ) {
+		if ( 0 === strcasecmp( $shortName, $fullName ) ) {
+			return true;
+		}
+		
+		if ( preg_match( "/^{$shortName}[- ]/i", $fullName ) ) {
+			return true;
+		}
+		
+		return false;
+	}
+	
+	/**
+	 * Transform matches into a member match object
+	 *
+	 * @param Member[] $matches
+	 * @param bool $ambiguous
+	 *
+	 * @return MemberMatch
+	 */
+	private static function create( array $matches, bool $ambiguous ): MemberMatch {
+		switch ( count( $matches ) ) {
+			case 0:
+				$status = self::NO_MATCH;
+				break;
+			
+			case 1:
+				$status = $ambiguous ? self::AMBIGUOUS_MATCH : self::MATCH;
+				break;
+			
+			default:
+				$status = self::MULTIPLE_MATCHES;
+		}
+		
+		// make sure the matches are strictly ordered (no holes from unset)
+		$matches = array_values( $matches );
+		
+		return new MemberMatch( $status, $matches );
+	}
+	
+	/**
+	 * Return webling query to find members by first and last name
+	 *
+	 * @param Member $member
+	 *
+	 * @return string
+	 */
+	private static function buildNameQuery( Member $member ): string {
+		$query = [];
+		if ( $member->firstName->getWeblingValue() ) {
+			$query[] = "`{$member->firstName->getWeblingKey()}` FILTER '{$member->firstName->getWeblingValue()}'";
+		}
+		
+		if ( $member->lastName->getWeblingValue() ) {
+			$query[] = "`{$member->lastName->getWeblingKey()}` FILTER '{$member->lastName->getWeblingValue()}'";
+		}
+		
+		return implode( ' AND ', $query );
+	}
+	
+	/**
+	 * Remove members where trailing characters of name are not separated by
+	 * either a space or a hyphen.
+	 *
+	 * There are two reasons for this function:
+	 * - In online forms, people tend to not write there full names (if they have multiple)
+	 * - Weblings FILTER is a 'starts with' function, that doesn't allow any further checks
+	 *
+	 * @param Member[] $matches
+	 * @param string $firstName
+	 * @param string $lastName
+	 */
+	private static function removeWrongNameMatches( array &$matches, string $firstName, string $lastName ) {
+		foreach ( $matches as $idx => $match ) {
+			if ( ! self::isShortNameOf( $firstName, $match->firstName->getValue() )
```

Comment:
Die erste Hälfte macht das gleiche wie `selectByFirstName`, oder?
Ich würde direkt selectByFirstName aufrufen, hat aber den Nachteil, dass man den foreach-loop 2x macht.

Reply:
bei `selectByFirstName` wird die Namensprüfung nur durchgeführt, wenn der `firstName` des `$match` nicht leer ist. Bei `removeWrongNameMatches` wird die Prüfung in jedem Fall durchgeführt, weil der `firstName` nicht leer sein darf/kann (sonst würde es sehr viele falsche matches geben).  Aus diesem Grund habe ich auch unterschiedliche Namen gewählt.

Reply:
`if ( $match->firstName->getValue() )` habe ich übersehen. Es würde in dem Fall einen zusätzlichen Parameter brauchen, um anzugeben, was bei leerem $firstName passieren soll. Aber so funktionierts ja auch. Macht also keinen Sinn, da gross umzubauen.
Ich bin vielleicht etwas ein Dont-Repeat-Yourself-Purist :smile:

Reply:
Ich finde es auch ein bisschen zwiespältig. Aber da es vom Sinn her ein anderer Check ist, habe ich mich gegen weitere Normalisierung entscheiden, da diese bei Änderungen nachher ungewollte Seiteneffekte haben kann. Aber ist durchaus kein eindeutiger Fall.